src/main/java/picocli/NameMatcher.java

```diff
@@ -0,0 +1,80 @@
+package picocli;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Set;
+
+class NameMatcher {
```

Comment:
I am not sure yet where these methods should live...

* they should be easy to test, so package-protected methods of a public or package-protected class
* they should not be in an external top-level class, but somewhere in the `CommandLine.java` source file
* they should not make the picocli jar bigger than necessary (unsure how much difference it makes in size if we add a new static inner class for these methods, but perhaps `ParserSpec` or `CommandLine` is a good place... still thinking...)

Reply:
This class will need information from `CommandSpec` for case-insensitivity, and if it is not in `CommandSpec`, it should be in `CommandLine` to be able to access the getter.

Reply:
Or we can (must) pass the case-insensitivity as an argument to the methods, since the case-insensitivity might be different for subcommands and options, while we only have one `match` method.

Reply:
Passing case-insensitivity as an argument the the `match` method makes sense to me.